File: mobile/ios/SundayDesk/UrlPolicy.swift
import Foundation

enum UrlPolicy {
    static let appHost = "fantasy-football-helper-orcin.vercel.app"
    static let appOrigin = "https://" + appHost
    static let setupURL = URL(string: appOrigin + "/setup")!
    static let espnURL = URL(string: "https://www.espn.com/fantasy/football/")!

    private static func secure(_ url: URL?) -> URLComponents? {
        guard let url, let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              components.scheme?.lowercased() == "https", components.host != nil,
              components.user == nil, components.password == nil,
              components.port == nil || components.port == 443 else { return nil }
        return components
    }

    static func isApp(_ url: URL?) -> Bool {
        guard let components = secure(url) else { return false }
        return components.host?.lowercased() == appHost
    }

    static func isSetup(_ url: URL?) -> Bool {
        guard let components = secure(url) else { return false }
        return components.host?.lowercased() == appHost && components.percentEncodedPath == "/setup"
    }

    static func isProvider(_ url: URL?) -> Bool {
        guard let components = secure(url), let host = components.host?.lowercased() else { return false }
        return ["espn.com", "go.com", "disney.com"].contains { inDomain(host, $0) }
    }

    static func inDomain(_ host: String, _ domain: String) -> Bool {
        host == domain || host.hasSuffix("." + domain)
    }

    static func isRequestId(_ id: String) -> Bool {
        id.count == 36 && UUID(uuidString: id) != nil
    }
}

struct EspnSession {
    let s2: String
    let swid: String

    static func isValid(_ s2: String, _ swid: String) -> Bool {
        (16...8192).contains(s2.count) && !s2.contains("\r") && !s2.contains("\n") && !s2.contains(";")
            && swid.count == 38 && swid.first == "{" && swid.last == "}"
            && UUID(uuidString: String(swid.dropFirst().dropLast())) != nil
    }
}

File: mobile/android/app/src/main/java/dev/acebuilds/sundaydesk/UrlPolicy.java
package dev.acebuilds.sundaydesk;

import java.net.URI;
import java.util.Locale;
import java.util.regex.Pattern;

/** Shared policy has no Android dependencies, so it can be unit tested on the JVM. */
final class UrlPolicy {
    static final String APP_ORIGIN = "https://fantasy-football-helper-orcin.vercel.app";
    static final String SETUP_URL = APP_ORIGIN + "/setup";
    static final String ESPN_URL = "https://www.espn.com/fantasy/football/";
    private static final String APP_HOST = "fantasy-football-helper-orcin.vercel.app";
    private static final Pattern UUID = Pattern.compile("^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$");

    private static URI secure(String value) {
        try {
            URI url = new URI(value);
            if (!"https".equalsIgnoreCase(url.getScheme()) || url.getHost() == null
                || url.getRawUserInfo() != null || (url.getPort() != -1 && url.getPort() != 443)) return null;
            return url;
        } catch (Exception ignored) { return null; }
    }

    static boolean isApp(String value) {
        URI url = secure(value);
        return url != null && APP_HOST.equalsIgnoreCase(url.getHost());
    }

    static boolean isSetup(String value) {
        URI url = secure(value);
        return url != null && APP_HOST.equalsIgnoreCase(url.getHost()) && "/setup".equals(url.getRawPath());
    }

    static boolean isProvider(String value) {
        URI url = secure(value);
        if (url == null) return false;
        String host = url.getHost().toLowerCase(Locale.ROOT);
        return inDomain(host, "espn.com") || inDomain(host, "go.com") || inDomain(host, "disney.com");
    }

    static boolean inDomain(String host, String domain) {
        return host.equals(domain) || host.endsWith("." + domain);
    }

    static boolean isRequestId(String value) { return value != null && UUID.matcher(value).matches(); }
}
